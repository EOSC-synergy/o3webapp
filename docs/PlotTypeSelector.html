<!DOCTYPE html>
<html lang="en">

<head>
  
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title> PlotTypeSelector</title>

  <script src="https://cdn.jsdelivr.net/gh/google/code-prettify@master/loader/run_prettify.js"></script>
  <script src="https://unpkg.com/@babel/standalone/babel.min.js"></script>
  <script src="./build/entry.js"></script>
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->
  <link href="https://fonts.googleapis.com/css?family=Roboto:100,400,700|Inconsolata,700" rel="stylesheet">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.3/css/all.css" integrity="sha384-UHRtZLI+pbxtHCWp1t77Bi1L4ZtiqrqD80Kn4Z8NTSRyMA2Fd33n5dQ8lWUE00s/" crossorigin="anonymous">
  <link type="text/css" rel="stylesheet" href="https://jmblog.github.io/color-themes-for-google-code-prettify/themes/tomorrow-night.min.css">
  <link type="text/css" rel="stylesheet" href="styles/app.min.css">
  <link type="text/css" rel="stylesheet" href="styles/iframe.css">
  <link type="text/css" rel="stylesheet" href="">
  <script async defer src="https://buttons.github.io/buttons.js"></script>

  
</head>



<body class="layout small-header">
    <div id="stickyNavbarOverlay"></div>
    

<div class="top-nav">
    <div class="inner">
        <a id="hamburger" role="button" class="navbar-burger" aria-label="menu" aria-expanded="false">
            <span aria-hidden="true"></span>
            <span aria-hidden="true"></span>
            <span aria-hidden="true"></span>
        </a>
        <div class="logo">
            
             
                <a href="index.html">
                    <h1 class="navbar-item">O3as Webapp documentation</h1>
                </a>
            
        </div>
        <div class="menu">
            
            <div class="navigation">
                <a
                    href="index.html"
                    class="link"
                >
                    Documentation
                </a>
                
                
                
            </div>
        </div>
    </div>
</div>
    <div id="main">
        <div
            class="sidebar "
            id="sidebarNav"
        >
            
            <nav>
                
                    <h2><a href="index.html">Documentation</a></h2><div class="category"><h3>Modules</h3><ul><li><a href="module-API.html">API</a></li><li><a href="module-ArrayOperations.html">ArrayOperations</a></li><li><a href="module-Constants.html">Constants</a></li><li><a href="module-CSV.html">CSV</a></li><li><a href="module-DownloadNotPdf.html">DownloadNotPdf</a></li><li><a href="module-Math.html">Math</a></li><li><a href="module-ModelNameConverter.html">ModelNameConverter</a></li><li><a href="module-OptionsFormatter.html">OptionsFormatter</a></li><li><a href="module-PdfCreator.html">PdfCreator</a></li><li><a href="module-TextSearch.html">TextSearch</a></li></ul><h3>Components</h3><ul><li><a href="AddModelGroupModal.html">AddModelGroupModal</a></li><li><a href="App.html">App</a></li><li><a href="CookieConsentModal.html">CookieConsentModal</a></li><li><a href="CustomLatitudeSelector.html">CustomLatitudeSelector</a></li><li><a href="DiscardChangesModal.html">DiscardChangesModal</a></li><li><a href="DownloadModal.html">DownloadModal</a></li><li><a href="EditModelGroupModal.html">EditModelGroupModal</a></li><li><a href="ErrorMessageModal.html">ErrorMessageModal</a></li><li><a href="Footer.html">Footer</a></li><li><a href="Graph.html">Graph</a></li><li><a href="LandingPage.html">LandingPage</a></li><li><a href="LatitudeBandSelector.html">LatitudeBandSelector</a></li><li><a href="Logo.html">Logo</a></li><li><a href="ModelGroupCard.html">ModelGroupCard</a></li><li><a href="ModelGroupConfigurator.html">ModelGroupConfigurator</a></li><li><a href="Navbar.html">Navbar</a></li><li><a href="PlotNameField.html">PlotNameField</a></li><li><a href="PlotTypeSelector.html">PlotTypeSelector</a></li><li><a href="ReferenceModelSelector.html">ReferenceModelSelector</a></li><li><a href="ReferenceYearField.html">ReferenceYearField</a></li><li><a href="RegionSelector.html">RegionSelector</a></li><li><a href="SearchBar.html">SearchBar</a></li><li><a href="SeasonCheckBoxGroup.html">SeasonCheckBoxGroup</a></li><li><a href="Section.html">Section</a></li><li><a href="Sidebar.html">Sidebar</a></li><li><a href="TimeCheckBoxGroup.html">TimeCheckBoxGroup</a></li><li><a href="XAxisField.html">XAxisField</a></li><li><a href="YAxisField.html">YAxisField</a></li></ul></div><div class="category"><h2>apiSlice</h2><h3>Global</h3><ul><li><a href="global.html#apiSlice">apiSlice</a></li><li><a href="global.html#extraReducers">extraReducers</a></li><li><a href="global.html#fetchModels">fetchModels</a></li><li><a href="global.html#fetchPlotData">fetchPlotData</a></li><li><a href="global.html#fetchPlotDataForCurrentModels">fetchPlotDataForCurrentModels</a></li><li><a href="global.html#fetchPlotDataInitiallyPending">fetchPlotDataInitiallyPending</a></li><li><a href="global.html#fetchPlotDataRefetching">fetchPlotDataRefetching</a></li><li><a href="global.html#fetchPlotDataRejected">fetchPlotDataRejected</a></li><li><a href="global.html#fetchPlotDataSuccess">fetchPlotDataSuccess</a></li><li><a href="global.html#fetchPlotTypes">fetchPlotTypes</a></li><li><a href="global.html#generateCacheKey">generateCacheKey</a></li><li><a href="global.html#getAllSelectedModels">getAllSelectedModels</a></li><li><a href="global.html#initialState">initialState</a></li><li><a href="global.html#REQUEST_STATE">REQUEST_STATE</a></li><li><a href="global.html#selectActivePlotData">selectActivePlotData</a></li><li><a href="global.html#selectExistingPlotData">selectExistingPlotData</a></li><li><a href="global.html#updateDataAndDisplaySuggestions">updateDataAndDisplaySuggestions</a></li></ul></div><div class="category"><h2>modelSlice</h2><h3>Global</h3><ul><li><a href="global.html"></a></li><li><a href="global.html#initialState">initialState</a></li><li><a href="global.html#MODEL_DATA_TEMPLATE">MODEL_DATA_TEMPLATE</a></li><li><a href="global.html#MODEL_GROUP_TEMPLATE">MODEL_GROUP_TEMPLATE</a></li><li><a href="global.html#modelsSlice">modelsSlice</a></li><li><a href="global.html#selectAllGroupIds">selectAllGroupIds</a></li><li><a href="global.html#selectAllModelGroups">selectAllModelGroups</a></li><li><a href="global.html#selectModelDataOfGroup">selectModelDataOfGroup</a></li><li><a href="global.html#selectModelsOfGroup">selectModelsOfGroup</a></li><li><a href="global.html#selectNameOfGroup">selectNameOfGroup</a></li><li><a href="global.html#selectStatisticalValueSettingsOfGroup">selectStatisticalValueSettingsOfGroup</a></li><li><a href="global.html#selectVisibilityOfGroup">selectVisibilityOfGroup</a></li></ul></div><div class="category"><h2>plotSlice</h2><h3>Global</h3><ul><li><a href="global.html"></a></li><li><a href="global.html#DEFAULT_MONTHS">DEFAULT_MONTHS</a></li><li><a href="global.html#initialState">initialState</a></li><li><a href="global.html#plotSlice">plotSlice</a></li><li><a href="global.html#selectPlotId">selectPlotId</a></li><li><a href="global.html#selectPlotLocation">selectPlotLocation</a></li><li><a href="global.html#selectPlotMonths">selectPlotMonths</a></li><li><a href="global.html#selectPlotTitle">selectPlotTitle</a></li><li><a href="global.html#selectPlotXRange">selectPlotXRange</a></li><li><a href="global.html#selectPlotYRange">selectPlotYRange</a></li></ul></div><div class="category"><h2>referenceSlice</h2><h3>Global</h3><ul><li><a href="global.html"></a></li><li><a href="global.html#DEFAULT_REF_MODEL">DEFAULT_REF_MODEL</a></li><li><a href="global.html#DEFAULT_REF_YEAR">DEFAULT_REF_YEAR</a></li><li><a href="global.html#initialState">initialState</a></li><li><a href="global.html#referenceSlice">referenceSlice</a></li><li><a href="global.html#selectIsOffsetApplied">selectIsOffsetApplied</a></li><li><a href="global.html#selectRefModel">selectRefModel</a></li><li><a href="global.html#selectRefYear">selectRefYear</a></li><li><a href="global.html#selectVisibility">selectVisibility</a></li></ul></div>
                
            </nav>
        </div>
        <div class="core" id="main-content-wrapper">
            <div class="content">
                <header class="page-title">
                    <p>Components</p>
                    <h1>PlotTypeSelector</h1>
                </header>
                




<section>

<header>
    
        <h2>
            &ltPlotTypeSelector <i>reportError</i> /&gt
        </h2>
        
    
</header>

<article>
    <div class="container-overview">
    
        
            <div class='vertical-section'>
                <div class="members">
                    <div class="member">
                        <div class=name>
                            <span class="tag">Constructor</span>
                        </div>
                        


    
    <h4 class="name" id="PlotTypeSelector">
        <a class="href-link" href="#PlotTypeSelector">#</a>
        
        <span class="code-name">
            
                &ltPlotTypeSelector <i>reportError</i> /&gt
            
        </span>
    </h4>
    

    
    
    <div class="description">
        <p>Enables the user to select a different plot type.</p>
    </div>
    







    
        <h5>PropTypes:</h5>
        
<div class="table-container">
  <table class="params table">
    <thead>
    <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Required</th>
        <th>Description</th>
        <th class="last">Default</th>
    </tr>
    </thead>

    <tbody>
    
          <tr class="deep-level-0">
            <td class="name"><code>reportError</code></td>
            <td>
              func
            </td>
            <td>Yes</td>
            <td>function for error handling</td>
            <td></td>
          </tr>
          
    
    </tbody>
  </table>
</div>
    



    <h5>Parameters:</h5>
    
<div class="table-container">
    <table class="params table">
        <thead>
        <tr>
            
            <th>Name</th>
            

            <th>Type</th>

            

            

            <th class="last">Description</th>
        </tr>
        </thead>

        <tbody>
        

            
<tr class="deep-level-0">
  
      <td class="name"><code>props</code></td>
  

  <td class="type">
  
      
<code class="param-type">Object</code>


  
  </td>

  

  

  <td class="description last"><p>specified by propTypes</p></td>
</tr>


        
        </tbody>
    </table>
</div>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html" class="button">View Source</a>
            <span>
                <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html">views/landingPage/Sidebar/InputComponents/PlotTypeSelector/PlotTypeSelector.js</a>, <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html#line20">line 20</a>
            </span>
        </p>
    
</dl>


















    <div class='columns method-parameter'>
        <div class="column is-2"><label>Returns:</label></div>
        <div class="column is-10">
            
                    

<div class="columns">
    
    <div class='param-desc column is-7'><p>a jsx containing a dropdown to select the plot type</p></div>
    
    
    <div class='column is-5 has-text-left'>
        <label>Type: </label>
        
<code class="param-type">JSX.Element</code>


    </div>
    
</div>

                
        </div>
    </div>





                    </div>
                </div>
            </div>
        
    
    </div>
    
        
    
    

    

    

    

    

    

    
        <div class='vertical-section'>
            <h1>Members</h1>
            <div class="members">
            
                <div class="member">

    <span class="method-parameter is-pulled-right">
        <label>Type:</label>
        
<code class="param-type">function</code>


    </span>

<h4 class="name" id="~dispatch">
    <a class="href-link" href="#~dispatch">#</a>
    
        
            <span class='tag'>inner</span>
        
            <span class='tag'>constant</span>
        
    
    <span class="code-name">
        dispatch
    </span>
    
</h4>




<div class="description">
    <p>A dispatch function to dispatch actions to the Redux store.</p>
</div>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html" class="button">View Source</a>
            <span>
                <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html">views/landingPage/Sidebar/InputComponents/PlotTypeSelector/PlotTypeSelector.js</a>, <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html#line26">line 26</a>
            </span>
        </p>
    
</dl>





</div>
            
                <div class="member">

    <span class="method-parameter is-pulled-right">
        <label>Type:</label>
        
<code class="param-type">String</code>


    </span>

<h4 class="name" id="~plotType">
    <a class="href-link" href="#~plotType">#</a>
    
        
            <span class='tag'>inner</span>
        
            <span class='tag'>constant</span>
        
    
    <span class="code-name">
        plotType
    </span>
    
</h4>




<div class="description">
    <p>Currently selected plot type. Taken from the redux store.</p>
</div>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    
    <dt class="tag-see">See:</dt>
    <dd class="tag-see">
        <ul>
            <li><a href="global.html#selectPlotId">selectPlotId</a></li>
        </ul>
    </dd>
    

    

    
        <p class="tag-source">
            <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html" class="button">View Source</a>
            <span>
                <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html">views/landingPage/Sidebar/InputComponents/PlotTypeSelector/PlotTypeSelector.js</a>, <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html#line39">line 39</a>
            </span>
        </p>
    
</dl>





</div>
            
                <div class="member">

    <span class="method-parameter is-pulled-right">
        <label>Type:</label>
        
<code class="param-type">Object</code>


    </span>

<h4 class="name" id="~plotTypesRequestData">
    <a class="href-link" href="#~plotTypesRequestData">#</a>
    
        
            <span class='tag'>inner</span>
        
            <span class='tag'>constant</span>
        
    
    <span class="code-name">
        plotTypesRequestData
    </span>
    
</h4>




<div class="description">
    <p>Get the requested data from the redux store</p>
</div>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html" class="button">View Source</a>
            <span>
                <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html">views/landingPage/Sidebar/InputComponents/PlotTypeSelector/PlotTypeSelector.js</a>, <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html#line32">line 32</a>
            </span>
        </p>
    
</dl>





</div>
            
            </div>
        </div>
    

    
        <div class='vertical-section'>
            <h1>Methods</h1>
            <div class="members">
            
                <div class="member">


    
    <h4 class="name" id="~changePlotType">
        <a class="href-link" href="#~changePlotType">#</a>
        
            
                <span class='tag'>inner</span>
            
        
        <span class="code-name">
            
                changePlotType<span class="signature">(event)</span><span class="type-signature"></span>
            
        </span>
    </h4>
    

    
    
    <div class="description">
        <p>Calls the redux store to change the plot type</p>
    </div>
    









    <h5>Parameters:</h5>
    
<div class="table-container">
    <table class="params table">
        <thead>
        <tr>
            
            <th>Name</th>
            

            <th>Type</th>

            

            

            <th class="last">Description</th>
        </tr>
        </thead>

        <tbody>
        

            
<tr class="deep-level-0">
  
      <td class="name"><code>event</code></td>
  

  <td class="type">
  
      
<code class="param-type">event</code>


  
  </td>

  

  

  <td class="description last"><p>the event that called this function</p></td>
</tr>


        
        </tbody>
    </table>
</div>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html" class="button">View Source</a>
            <span>
                <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html">views/landingPage/Sidebar/InputComponents/PlotTypeSelector/PlotTypeSelector.js</a>, <a href="views_landingPage_Sidebar_InputComponents_PlotTypeSelector_PlotTypeSelector.js.html#line46">line 46</a>
            </span>
        </p>
    
</dl>





















</div>
            
            </div>
        </div>
    

    

    
</article>

</section>




            </div>
            
            <footer class="footer">
                <div class="content has-text-centered">
                    <p>Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.6.10</a></p>
                    <p class="sidebar-created-by">
                        <a href="https://github.com/SoftwareBrothers/better-docs" target="_blank">BetterDocs theme</a> provided with <i class="fas fa-heart"></i> by
                        <a href="http://softwarebrothers.co" target="_blank">SoftwareBrothers - JavaScript Development Agency</a>
                    </p>
                </div>
            </footer>
            
        </div>
        <div id="side-nav" class="side-nav">
        </div>
    </div>
<script src="scripts/app.min.js"></script>
<script>PR.prettyPrint();</script>
<script src="scripts/linenumber.js"> </script>


</body>
</html>