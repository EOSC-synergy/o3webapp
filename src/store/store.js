import { applyMiddleware, combineReducers, configureStore, createStore } from '@reduxjs/toolkit';
import plotReducer from './plotSlice/plotSlice';
import modelsReducer, { setModelsOfModelGroup } from './modelsSlice/modelsSlice';
import referenceReducer from './referenceSlice/referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { HYDRATE, createWrapper } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';

const bindMiddleware = (middleware) => {
    if (process.env.NODE_ENV !== 'production') {
        const { composeWithDevTools } = require('redux-devtools-extension');
        return composeWithDevTools(applyMiddleware(...middleware));
    }
    return applyMiddleware(...middleware);
};

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state, action) => {
    if (action.type === HYDRATE) {
        const nextState = {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return nextState;
    } else {
        return combinedReducer(state, action);
    }
};

let store;

const initStore = () => {
    store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return store;
};

export { store };

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = configureStore({
        reducer: combinedReducer,
    });
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
