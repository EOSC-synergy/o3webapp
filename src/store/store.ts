import { Action, combineReducers, configureStore, ThunkAction } from '@reduxjs/toolkit';
import plotReducer, { GlobalPlotState } from './plotSlice';
import modelsReducer, { GlobalModelState, setModelsOfModelGroup } from './modelsSlice';
import referenceReducer, { GlobalReferenceState } from './referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { createWrapper, HYDRATE } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';
import { useDispatch, useStore } from 'react-redux';

type GlobalState =
    | undefined
    | (GlobalPlotState & GlobalModelState & GlobalReferenceState & { api: any });

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state: GlobalState, action: { type: string; payload: any }) => {
    if (action.type === HYDRATE) {
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
    } else {
        return combinedReducer(state, action);
    }
};

//let store: ReturnType<typeof configureStore<GlobalState, >>;

const initStore = () => {
    //store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return configureStore({
        reducer,
        devTools: process.env.NODE_ENV !== 'production',
        middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
    });
};

export type AppStore = ReturnType<typeof initStore>;
export type AppState = ReturnType<AppStore['getState']>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>;

export type AppDispatch = AppStore['dispatch'];
export const useAppDispatch = () => useDispatch<AppDispatch>(); // Export a hook that can be reused to resolve types

export const useAppStore = () => useStore<AppState>();

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = initStore();
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
