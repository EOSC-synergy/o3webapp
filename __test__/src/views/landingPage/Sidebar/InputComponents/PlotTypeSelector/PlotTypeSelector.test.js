import { REQUEST_STATE } from '../../../../../../../src/services/API/apiSlice/apiSlice';
import React from 'react';
import ReactDOM from 'react-dom';
import PlotTypeSelector from '../../../../../../../src/views/landingPage/Sidebar/InputComponents/PlotTypeSelector';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { act } from 'react-dom/test-utils';
import { createTestStore } from '../../../../../../../src/store/store';
import { Provider } from 'react-redux';

let store;
describe('plot type selector test', () => {
    beforeEach(() => {
        store = createTestStore();
    });

    it('PlotTypeSelector renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <Provider store={store}>
                <PlotTypeSelector reportError={() => {}} />
            </Provider>,
            div
        );
    });

    // Snapshot test
    /*it('PlotTypeSelector renders correctly', () => {
        const { asFragment } = render(
            <Provider store={store}>
                <PlotTypeSelector reportError={() => {}} />
            </Provider>
        );
        expect(asFragment()).toMatchSnapshot();
        userEvent.tab(); // focuses select
        expect(asFragment()).toMatchSnapshot();
    });*/

    // Prop types
    it('PlotTypeSelector expects a report error function', () => {
        console.error = jest.fn();
        render(
            <Provider store={store}>
                <PlotTypeSelector />
            </Provider>
        );
        expect(console.error).toHaveBeenCalledTimes(1);
    });

    it('PlotTypeSelector reports error if an error occurred in the fetching of plotTypes', () => {
        const errorMessage = 'API not responding: Error Message';
        const mockDispatch = jest.fn();
        const mockSelector = jest.fn();
        jest.mock('react-redux', () => ({
            ...jest.requireActual('react-redux'),
            useDispatch: () => mockDispatch,
            useSelector: () => mockSelector,
        }));

        mockSelector.mockReturnValue({
            status: REQUEST_STATE.error,
            data: [],
            error: 'Error Message',
        });

        const reportError = jest.fn();
        const div = document.createElement('div');

        ReactDOM.render(
            <Provider store={store}>
                <PlotTypeSelector reportError={reportError} />
            </Provider>,
            div
        );
        expect(reportError.mock.calls[0][0]).toEqual(errorMessage);
    });

    it('should render all the returned data from the api', () => {
        const RETURNED_OPTIONS = ['tco3_zm', 'tco3_return'];

        const mockDispatch = jest.fn();
        const mockSelector = jest.fn();
        jest.mock('react-redux', () => ({
            ...jest.requireActual('react-redux'),
            useDispatch: () => mockDispatch,
            useSelector: () => mockSelector,
        }));

        mockSelector
            .mockReturnValueOnce(
                // mock api request
                {
                    status: REQUEST_STATE.success,
                    data: RETURNED_OPTIONS,
                    error: null,
                }
            )
            .mockReturnValueOnce(RETURNED_OPTIONS[0]);

        const { getByRole, getAllByRole } = render(
            <Provider store={store}>
                <PlotTypeSelector reportError={() => {}} />
            </Provider>
        );

        const trigger = getByRole('button');
        fireEvent.mouseDown(trigger);

        const options = getAllByRole('option');

        console.debug(options, RETURNED_OPTIONS);

        expect(options.length).toEqual(RETURNED_OPTIONS.length);
        for (let i; i < options.length; ++i) {
            expect(options[i].textContent).toEqual(RETURNED_OPTIONS[i]);
        }
    });

    it('should update the clicked value in the store', () => {
        const RETURNED_OPTIONS = ['tco3_zm', 'tco3_return'];

        const mockDispatch = jest.fn();
        const mockSelector = jest.fn();
        jest.mock('react-redux', () => ({
            ...jest.requireActual('react-redux'),
            useDispatch: () => mockDispatch,
            useSelector: () => mockSelector,
        }));

        mockSelector
            .mockReturnValueOnce(
                // mock api request
                {
                    status: REQUEST_STATE.success,
                    data: RETURNED_OPTIONS,
                    error: null,
                }
            )
            .mockReturnValueOnce(RETURNED_OPTIONS[0]);

        expect(store.getState().plot.plotId).toEqual(RETURNED_OPTIONS[0]); // default in store

        const { getByRole, getAllByRole } = render(
            <Provider store={store}>
                <PlotTypeSelector reportError={() => {}} />
            </Provider>
        );

        const trigger = getByRole('button');
        fireEvent.mouseDown(trigger);

        const options = getAllByRole('option');

        act(() => {
            options[1].click(); // click on second option (tco3_return)
        });
        expect(store.getState().plot.plotId).toEqual(RETURNED_OPTIONS[1]); // after selection
    });

    it('should display a spinner while loading', () => {
        const mockDispatch = jest.fn();
        const mockSelector = jest.fn();
        jest.mock('react-redux', () => ({
            ...jest.requireActual('react-redux'),
            useDispatch: () => mockDispatch,
            useSelector: () => mockSelector,
        }));

        mockSelector
            .mockReturnValueOnce(
                // mock api request
                {
                    status: REQUEST_STATE.loading,
                    data: [],
                    error: null,
                }
            )
            .mockReturnValueOnce('tco3_zm');

        const { getByRole } = render(
            <Provider store={store}>
                <PlotTypeSelector reportError={() => {}} />
            </Provider>
        );

        const trigger = getByRole('button');
        fireEvent.mouseDown(trigger);

        const loader = screen.getByTestId('plotTypeSelectorLoading');
        expect(loader).toBeInTheDocument();
    });
});
