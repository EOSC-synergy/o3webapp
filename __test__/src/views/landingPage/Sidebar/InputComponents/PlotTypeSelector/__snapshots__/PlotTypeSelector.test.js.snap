// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plot type selector test PlotTypeSelector renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root css-1y0pph0-MuiFormControl-root"
    data-testid="plotTypeSelectorForm"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      data-testid="plotTypeSelector"
      id="plotTypeLabel"
    >
      Plot Type
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="plotTypeLabel plotType"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="plotType"
        role="button"
        tabindex="0"
      >
        <span
          class="notranslate"
        >
          ​
        </span>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1ftyaf0"
        >
          <span>
            Plot Type
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;

exports[`plot type selector test PlotTypeSelector renders correctly 2`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root css-1y0pph0-MuiFormControl-root"
    data-testid="plotTypeSelectorForm"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      data-testid="plotTypeSelector"
      id="plotTypeLabel"
    >
      Plot Type
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="plotTypeLabel plotType"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="plotType"
        role="button"
        tabindex="0"
      >
        <span
          class="notranslate"
        >
          ​
        </span>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1ftyaf0"
        >
          <span>
            Plot Type
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
